src/app/labeling/qgsmaptoolmovelabel.cpp

```diff
@@ -58,17 +62,54 @@ void QgsMapToolMoveLabel::deleteRubberBands()
 
 void QgsMapToolMoveLabel::cadCanvasMoveEvent( QgsMapMouseEvent *e )
 {
+
   if ( mLabelRubberBand )
   {
     const QgsPointXY pointMapCoords = e->mapPoint();
-    const double offsetX = pointMapCoords.x() - mStartPointMapCoords.x();
-    const double offsetY = pointMapCoords.y() - mStartPointMapCoords.y();
-    mLabelRubberBand->setTranslationOffset( offsetX, offsetY );
-    mLabelRubberBand->updatePosition();
-    mLabelRubberBand->update();
-    mFixPointRubberBand->setTranslationOffset( offsetX, offsetY );
-    mFixPointRubberBand->updatePosition();
-    mFixPointRubberBand->update();
+
+    bool isCurvedOrLine { mCurrentLabel.settings.placement == QgsPalLayerSettings::Placement::Curved || mCurrentLabel.settings.placement == QgsPalLayerSettings::Placement::Line };
+    if ( isCurvedOrLine )
+    {
+      // Determine the closest point on the feature
+      const QgsFeatureId featureId = mCurrentLabel.pos.featureId;
+      if ( mCurrentLabel.layer )
+      {
+        const QgsFeature feature { mCurrentLabel.layer->getFeature( featureId ) };
+        const QgsMapSettings &ms = mCanvas->mapSettings();
+        const QgsPointXY pointMapTransformed = ms.mapToLayerCoordinates( mCurrentLabel.layer, pointMapCoords );
+        const QgsGeometry pointMapGeometry = QgsGeometry::fromPointXY( pointMapTransformed );
+        if ( feature.geometry().distance( QgsGeometry::fromPointXY( pointMapCoords ) ) / mCanvas->mapUnitsPerPixel() > 100.0 )
```

Comment:
We'd need to transform the feature's geometry to the canvas CRS before this line